src/Collections/Sequences/NatSeq.dfy

```diff
@@ -0,0 +1,602 @@
+// RUN: %dafny /compile:0 /noNLarith "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*******************************************************************************
+*  Original: Copyright (c) 2020 Secure Foundations Lab
+*  SPDX-License-Identifier: MIT
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+/* The first element of a sequence is the least significant word; the last
+element is the most significant word. */
+
+include "../../NonlinearArithmetic/DivMod.dfy"
+include "../../NonlinearArithmetic/Mul.dfy"
+include "../../NonlinearArithmetic/Power.dfy"
+include "Seq.dfy"
+
+abstract module NatSeq {
+
+  import opened DivMod
+  import opened Mul
+  import opened Power
+  import opened Seq
+
+  /* Upper bound of a word */
+  function method BOUND(): nat
+    ensures BOUND() > 1
+
+  type uint = i: int | 0 <= i < BOUND()
+
+  //////////////////////////////////////////////////////////////////////////////
+  //
+  // ToNat definition and lemmas
+  //
+  //////////////////////////////////////////////////////////////////////////////
+
+  /* Converts a sequence to a nat beginning from the least significant word. */
```

Comment:
I just realized that this is a little-endian interpretation rather than big-endian. I'm not surprised given this code was used in verifying chipsets, but I worry that in a general-purpose library context, most readers will assume sequences are big-endian. That's the natural way to read them in source code: `[1, 5, 3]` with a base of `10` is very naturally interpreted as `153` since that matches how you'd read the literal `153` value.

Would switching this and having to sprinkle lots of `Seq.Reverse` calls in consumers that need little-endianness make it very difficult for those consumers? Could we add an example of using these modules in both endian modes either way?

Reply:
Thanks to Intel, we primarily live in a little-endian world, so at least in all of our systems-oriented projects, that's been the dominant mode for reasoning about, say, casts between uint32s and bytes.  We can certainly be more explicit about the fact that the default is little-endian, and we could perhaps expand to offer big-endian variants as well.

Reply:
This suggests a better name to me, actually: `LittleEndianNat`. That implies it's a positional number anyway, and leaves the door open for a `BigEndianNat`.

Reply:
Ah, yes, that would make it much clearer.  Good idea.